Fix inaccurate dispatch tables

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #53973 Add some trivial meta functions for commonly used inplace operations
* **#54127 Fix inaccurate dispatch tables**
* #54079 Add Tensor::is_cpu, genericize TensorIterator
* #54034 Refactor tensor_new.cpp to use TensorOptions instead of DispatchKey

During the meta tensor bringup, I found all of these operators
advertised that they worked on all backends (DefaultBackend/Math)
but actually they only worked on CPU/CUDA.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D27109508](https://our.internmc.facebook.com/intern/diff/D27109508)

ezyang committed:
Fix inaccurate dispatch tables

During the meta tensor bringup, I found all of these operators
advertised that they worked on all backends (DefaultBackend/Math)
but actually they only worked on CPU/CUDA.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Fix inaccurate dispatch tables"


During the meta tensor bringup, I found all of these operators
advertised that they worked on all backends (DefaultBackend/Math)
but actually they only worked on CPU/CUDA.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D27109508](https://our.internmc.facebook.com/intern/diff/D27109508)

[ghstack-poisoned]

ezyang committed:
Update on "Fix inaccurate dispatch tables"


During the meta tensor bringup, I found all of these operators
advertised that they worked on all backends (DefaultBackend/Math)
but actually they only worked on CPU/CUDA.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D27109508](https://our.internmc.facebook.com/intern/diff/D27109508)

[ghstack-poisoned]

ezyang committed:
Update on "Fix inaccurate dispatch tables"


During the meta tensor bringup, I found all of these operators
advertised that they worked on all backends (DefaultBackend/Math)
but actually they only worked on CPU/CUDA.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D27109508](https://our.internmc.facebook.com/intern/diff/D27109508)

[ghstack-poisoned]

mruberry reviewed (approved):
Nice cleanup. As mentioned on another PR:

- is there a way to test for this so new incorrect dispatches can't accumulate?
- are there clearer names we can use than "DefaultBackend" or "Math"? "DefaultBackend" seems to be especially confusing.

ezyang committed:
Update on "Fix inaccurate dispatch tables"


During the meta tensor bringup, I found all of these operators
advertised that they worked on all backends (DefaultBackend/Math)
but actually they only worked on CPU/CUDA.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D27109508](https://our.internmc.facebook.com/intern/diff/D27109508)

[ghstack-poisoned]

ailzhang reviewed (approved):
Thanks!
